### ISSUE TITLE
Pyzo takes 2 seconds to close

### ISSUE THREAD

**Issue by l33tlinuxh4x0r** [pyzo/pyzo#707]
Pyzo hangs on close for about 2 seconds then closes. Every other program/app on my computer closes instantly when I press The X button or exit from the file menu.

I have tested this on Arch Linux and Windows 10. I'm not sure what is causing this.

**Comment by almarklein** [comment]
Is this the latest version?

**Comment by l33tlinuxh4x0r** [comment]
yes

**Comment by almarklein** [comment]
Pyzo does do some cleanup when closing, which is mostly making sure that the kernels all exit. Do the kernels (shells) restart quickly when you hit that restart button (or ctrl+k) in the shell?

**Comment by l33tlinuxh4x0r** [comment]
Yes very quickly.

The hang on close is almost as if the timer for the splash screen that says pyzo is rerun before pyzo closes. It is almost the exact same amount of time that pyzo hangs. Also is there a way to disable the splash screen in settings?

**Comment by almarklein** [comment]
I can actually confirm that it takes 1 or 2 seconds to close Pyzo. I think you can consider this normal behavior. It does a few cleanup tasks, like closing the shell subprocess, checking that all open files have no unsaved changes, stopping threads etc. It could be that one of these tasks takes more time than is needed. Someone could do some digging by doing time measurements on these steps. 

I'll be leaving this issue open to track this task. If anyone is interested, the code in question is [here](https://github.com/pyzo/pyzo/blob/master/pyzo/core/main.py#L349-L390).

**Comment by almarklein** [comment]
> Also is there a way to disable the splash screen in settings?

No. Pyzo actually does some boot-stuff during the time that the splash is shown. Though it does make sure to show the splash screen for at least a second. We could perhaps remove that minimal time logic ...

**Comment by l33tlinuxh4x0r** [comment]
> I can actually confirm that it takes 1 or 2 seconds to close Pyzo. I think you can consider this normal behavior. It does a few cleanup tasks, like closing the shell subprocess, checking that all open files have no unsaved changes, stopping threads etc. It could be that one of these tasks takes more time than is needed. Someone could do some digging by doing time measurements on these steps.
> 
> I'll be leaving this issue open to track this task. If anyone is interested, the code in question is [here](https://github.com/pyzo/pyzo/blob/master/pyzo/core/main.py#L349-L390).

Just had a quick look at that code... What is up with the 0.1 on line 388? Does it add a 10th of a second delay before closing the next thread? I will try to download the source code and see if this is the problem.

**Comment by l33tlinuxh4x0r** [comment]
I have been commenting and un-commenting the lines of code that you posted. Looks like the problem is with the following lines of code.
```        # Proceed with closing shells
        pyzo.localKernelManager.terminateAll()
        for shell in pyzo.shells:
            shell._context.close()

        # Close tools
        for toolname in pyzo.toolManager.getLoadedTools():
            tool = pyzo.toolManager.getTool(toolname)
            tool.close()
```
With those lines commented out pyzo closes almost instantly like most other programs. I'm going to test and see if it leaves stuff in memory or running in the background without these lines. My guess is that when qt closes that it cleans up for us. I will edit this post after investigating further.

**Comment by almarklein** [comment]
I was away for a few days, thus a somewhat late reply.

I dug into these. The `pyzo.localKernelManager.terminateAll()` takes a long time, and I found out why. It terminates the kernels, but it does not get the right feedback (because of event loop stuff). This is fixed in #710, bringing it from 1 s to about 0.1 s.
The time for closing the shell contexts is negligible.

Some tools actually do need a cleanup. But I made that more explicit so we don't have to do it via the closeEvent. Plus there was another timing issue further down. Also fixed in #710.
